lib/Sema/TreeTransform.h

```diff
@@ -5228,11 +5230,47 @@ bool TreeTransform<Derived>::TransformFunctionTypeParams(
   return false;
 }
 
+template<typename Derived>
+bool TreeTransform<Derived>::TransformBoundsAnnotations(
+  BoundsAnnotations &Annot, bool &Changed) {
+  BoundsExpr *ExistingBounds = Annot.getBoundsExpr();
+  BoundsExpr *NewBounds = ExistingBounds;
+  if (ExistingBounds) {
+    ExprResult Result = getDerived().TransformExpr(ExistingBounds);
+    if (Result.isInvalid())
+      return true;
+    NewBounds = dyn_cast<BoundsExpr>(Result.get());
+    if (!NewBounds) {
+      assert("unexpected dynamic cast failure");
```

Comment:
Should this be something like `llvm_unreachable`? Is this a recoverable error, or something you never expect to see, since asserts will be compiled out in release mode.

Reply:
Good catch.  No, this isn't a recoverable error.  I'll change this and the other assert to be unreachable.